File: problems/贪心/452.用最少数量的箭引爆气球.py
#!/usr/bin/env python3
"""
452. 用最少数量的箭引爆气球
算法: 贪心（按右端点排序）
时间复杂度: O(n log n)
空间复杂度: O(1)
"""
def find_min_arrow_shots(points: list[list[int]]) -> int:
    if not points:
        return 0
    points.sort(key=lambda x: x[1])
    arrows = 1
    end = points[0][1]
    for i in range(1, len(points)):
        if points[i][0] > end:
            arrows += 1
            end = points[i][1]
    return arrows


if __name__ == "__main__":
    print(find_min_arrow_shots([[10, 16], [2, 8], [1, 6], [7, 12]]))  # 预期: 2
    print(find_min_arrow_shots([[1, 2], [3, 4], [5, 6]]))             # 预期: 3
    print(find_min_arrow_shots([[1, 2], [2, 3]]))                    # 预期: 1

File: problems/贪心/452.用最少数量的箭引爆气球.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int findMinArrowShots(vector<vector<int>>& points) {
    if (points.empty()) return 0;
    sort(points.begin(), points.end(),
         [](auto& a, auto& b) { return a[1] < b[1]; });
    int arrows = 1, end = points[0][1];
    for (int i = 1; i < points.size(); i++) {
        if (points[i][0] > end) {
            arrows++;
            end = points[i][1];
        }
    }
    return arrows;
}

int main() {
    // 测试用例
    vector<vector<int>> p1 = {{10, 16}, {2, 8}, {1, 6}, {7, 12}};
    cout << findMinArrowShots(p1) << endl;  // 预期: 2

    vector<vector<int>> p2 = {{1, 2}, {3, 4}, {5, 6}};
    cout << findMinArrowShots(p2) << endl;  // 预期: 3

    vector<vector<int>> p3 = {{1, 2}, {2, 3}};
    cout << findMinArrowShots(p3) << endl;  // 预期: 1

    return 0;
}
